src/test/java/com/google/sps/ContextServletTest.java

```diff
@@ -36,25 +45,44 @@
 public class ContextServletTest {
   private static final String URL_PARAMETER = "url";
   private static final String TEST_URL = "WkZ5e94QnWk";
-  private static final String REQUEST_INFO = "snippet,contentDetails,statistics";
+  private static final String TEST_TITLE = "Test Title";
+  private static final String TEST_CHANNEL = "Test Channel";
+  private static final int NUM_LIKES_VAL = 10;
+  private static final int NUM_DISLIKES_VAL = 1;
 
   @Test
-  public void testContextGet() throws ServletException, IOException, GeneralSecurityException {
-    ContextServlet contextServlet = new ContextServlet();
+  public void testContextGet() throws IOException, GeneralSecurityException {
+    // Test the doGet() method to retrieve youtube vidoe context with mocked service
+    ContextServlet contextServlet = spy(ContextServlet.class);
     HttpServletRequest mockedRequest = mock(HttpServletRequest.class);
     HttpServletResponse mockedResponse = mock(HttpServletResponse.class);
-    when(mockedRequest.getParameter(URL_PARAMETER)).thenReturn(TEST_URL);
     StringWriter stringWriter = new StringWriter();
     PrintWriter writer = new PrintWriter(stringWriter);
+    VideoSnippet mockedVideoSnippet =
+        new VideoSnippet()
+            .setPublishedAt(new DateTime(new Date()))
+            .setTitle(TEST_TITLE)
+            .setChannelTitle(TEST_CHANNEL);
+    VideoStatistics mockedVidoeStatistics =
+        new VideoStatistics().setLikeCount(BigInteger.TEN).setDislikeCount(BigInteger.ONE);
+    Video mockedVideo =
+        new Video().setSnippet(mockedVideoSnippet).setStatistics(mockedVidoeStatistics);
+    VideoListResponse mockedVideoList =
+        new VideoListResponse().setItems(Collections.singletonList(mockedVideo));
+    when(mockedRequest.getParameter(URL_PARAMETER)).thenReturn(TEST_URL);
+    when(contextServlet.generateYouTubeRequest(TEST_URL)).thenReturn(mockedVideoList);
```

Comment:
It's slightly strange to mock the `generateYoutubeRequest` method here since it means that we're mocking some of our own logic.

Ideally, the goal of the test should be two-fold:
1. Given a URL input to the servlet, a request to the Youtube API should be constructed as expected.
2. Given a response from the Youtube API, a JSON response is constructed as expected.

You've done a great job to accomplish 2, but we're currently missing some coverage for 1. 

Maybe we can talk offline to figure out the best to do 1? One thing that may make tesing this easier is invoking the `generateYoutubeRequest` directly in your test (instead of going through the servlet path).

Reply:
Resolved in 1-1